openssl/src/pkey_ctx.rs

```diff
@@ -0,0 +1,384 @@
+//! PKeyContext operations
+use crate::error::ErrorStack;
+#[cfg(ossl110)]
+use crate::hash::MessageDigest;
+use crate::pkey::{HasPrivate, HasPublic, PKeyRef};
+use crate::{cvt, cvt_p};
+use core::ptr;
+use foreign_types::ForeignTypeRef;
+use std::marker::PhantomData;
+#[cfg(ossl110)]
+use std::os::raw::{c_int, c_uchar};
+
+/// HKDF modes of operation. See [`hkdf_mode`](PKeyContext::hkdf_mode)
+#[cfg(ossl111)]
+pub enum HkdfMode {
+    /// Extract followed by expand
+    ExtractThenExpand = ffi::EVP_PKEY_HKDEF_MODE_EXTRACT_AND_EXPAND as isize,
+    /// Extract only HKDF
+    Extract = ffi::EVP_PKEY_HKDEF_MODE_EXTRACT_ONLY as isize,
+    /// Expand only HKDF
+    Expand = ffi::EVP_PKEY_HKDEF_MODE_EXPAND_ONLY as isize,
+}
+
+/// A type used for performing operations with PKeys
+pub struct PKeyContext<'a>(*mut ffi::EVP_PKEY_CTX, PhantomData<&'a ()>);
+
+unsafe impl<'a> Sync for PKeyContext<'a> {}
+unsafe impl<'a> Send for PKeyContext<'a> {}
+
+#[allow(clippy::len_without_is_empty)]
+impl<'a> PKeyContext<'a> {
+    /// Creates a new `PKeyContext` using the provided private key.
+    ///
+    /// This corresponds to [`EVP_PKEY_derive_init`].
+    ///
+    /// [`EVP_PKEY_derive_init`]: https://www.openssl.org/docs/man1.0.2/crypto/EVP_PKEY_derive_init.html
+    pub fn new_pkey<T>(key: &'a PKeyRef<T>) -> Result<PKeyContext<'a>, ErrorStack>
+    where
+        T: HasPrivate,
+    {
+        unsafe {
+            cvt_p(ffi::EVP_PKEY_CTX_new(key.as_ptr(), ptr::null_mut()))
+                .map(|p| PKeyContext(p, PhantomData))
+                .and_then(|ctx| cvt(ffi::EVP_PKEY_derive_init(ctx.0)).map(|_| ctx))
+        }
+    }
+
+    /// Sets the peer key used for secret derivation.
+    ///
+    /// This corresponds to [`EVP_PKEY_derive_set_peer`]:
+    ///
+    /// [`EVP_PKEY_derive_set_peer`]: https://www.openssl.org/docs/man1.0.2/crypto/EVP_PKEY_derive_init.html
+    pub fn set_peer<T>(&mut self, key: &'a PKeyRef<T>) -> Result<(), ErrorStack>
+    where
+        T: HasPublic,
+    {
+        unsafe { cvt(ffi::EVP_PKEY_derive_set_peer(self.0, key.as_ptr())).map(|_| ()) }
+    }
+
+    /// Creates a new `PKeyContext` in HKDF derivation mode using the digest provided.
+    ///
+    /// This corresponds to [`EVP_PKEY_CTX_new_id`] using `EVP_PKEY_HKDF` followed by
+    /// [`EVP_PKEY_derive_init`].
+    ///
+    /// [`EVP_PKEY_CTX_new_id`]: https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_CTX_new_id.html
+    /// [`EVP_PKEY_derive_init`]: https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_derive_init.html
+    #[cfg(ossl110)]
+    pub fn new_hkdf() -> Result<Self, ErrorStack> {
```

Comment:
Rather than having this specific to hkdf, I'd instead have a constructor taking `pkey::Id`.

Reply:
turned into `new_id` which requires an Id and the user will manually call derive_init after